cf-custom-resources/lib/nlb-cert-validator-updater.js

```diff
@@ -0,0 +1,311 @@
+                if (res.statusCode >= 400) {
+                    reject(new Error(`Error ${res.statusCode}: ${res.statusMessage}`));
+                } else {
+                    resolve();
+                }
+            })
+            .end(responseBody, "utf8");
+    });
+}
+
+exports.handler = async function (event, context) {
+    const props = event.ResourceProperties;
+
+    let {LoadBalancerDNS: loadBalancerDNS,
+        LoadBalancerHostedZoneID: loadBalancerHostedZoneID,
+        ServiceName: serviceName,
+        DomainName: domainName,
+    } = props;
+    let aliases = new Set(props.Aliases);
+
+    acm = new AWS.ACM();
+    envRoute53 = new AWS.Route53();
+    envHostedZoneID = props.EnvHostedZoneId;
+    envName = props.EnvName;
+    appName = props.AppName;
+    certificateDomain = `${serviceName}-nlb.${envName}.${appName}.${domainName}`;
+
+    let aliasesSorted = [...aliases].sort().join(",");
+    const physicalResourceID = `/${serviceName}/${aliasesSorted}`;
+
+    let handler = async function() {
+        switch (event.RequestType) {
+            case "Create":
+                await validateAliases(aliases, loadBalancerDNS);
+                const certificateARN = await requestCertificate(aliases, physicalResourceID, serviceName);
+                const options = await waitForValidationOptionsToBeReady(certificateARN, aliases);
+                await activate(options, certificateARN, loadBalancerDNS, loadBalancerHostedZoneID);
+                break;
+            case "Update":
+            case "Delete":
+            default:
+                throw new Error(`Unsupported request type ${event.RequestType}`);
+        }
+    };
+
+    try {
+        await handler();
+        await report(event, context, "SUCCESS", physicalResourceID);
+    } catch (err) {
+        console.log(`Caught error for service ${serviceName}: ${err.message}`);
+        await report(event, context, "FAILED", physicalResourceID, null, err.message);
+    }
+};
+
+/**
+ * Validate that the aliases are not in use.
+ *
+ * @param {Set<String>} aliases for the service.
+ * @param {String} loadBalancerDNS the DNS of the service's load balancer.
+ * @throws error if at least one of the aliases is not valid.
+ */
+async function validateAliases(aliases, loadBalancerDNS) {
+    let promises = [];
+
+    for (let alias of aliases) {
+        const promise = envRoute53.listResourceRecordSets({
+            HostedZoneId: envHostedZoneID,
+            MaxItems: "1",
+            StartRecordName: alias,
+        }).promise().then((data) => {
+            let recordSet = data["ResourceRecordSets"];
+            if (!recordSet || recordSet.length === 0) {
+                return;
+            }
+            if (recordSet[0].AliasTarget && recordSet[0].AliasTarget.DNSName === `${loadBalancerDNS}.`) {
+                return; // The record is an alias record and is in use by myself, hence valid.
+            }
+            throw new Error(`alias ${alias} is in use`);
```

Comment:
can we also add additional info for which LB uses it instead of the service's? something like:
```suggestion
            throw new Error(`Alias '${alias}' is already in use by another load balancer '${recordSet[0].AliasTarget.DNSName}'`);
```

Reply:
I will update the error message to indicate that the alias "could" be another load balancer if the alias target isn't null.

It is still possible that the domain is in use because people manually added a, e.g. CNAME, record using the domain routing to an IP address or another domain (not necessarily an alias to another service).